liabru/matter-js
# Timing, frame-rate consistency improvements and features

### Features

- improves all round consistency of the engine with respect to timing and frame-rates (delta)
- adds several new velocity and speed functions to `Matter.Body`
- adds `updateVelocity` argument to `Body.setPosition` etc. 
for e.g. moving bodies around directly but with imparted velocity
- changes body velocity and speed values to be frame-rate independent (via getters and setters)
- changes examples to be frame-rate independent
- removes `correction` from `Engine.update` as it is now built in

### Notes

For most purposes, these changes are __intended to be backwards compatible__. 

If you're using the engine at the standard browser `60fps` (`16.666ms` delta) then results should be essentially the same as before (or better) but __if they are not, please leave a comment and an example here if possible.__

If you're using `Body.setVelocity`, `Body.setAngularVelocity` at something _other than_ `60fps` (`16.666ms` delta) then you _may need to change your velocity values_ or remove any code you may have been using that factored in frame-rate to the velocity values, since velocity is now constant with respect to delta as expected.

While `body.velocity`  and `body.speed` remain the same, they are internal and not expected for use going forward, instead you should now use `Body.getVelocity` and `Body.getSpeed` which are frame-rate independent. Code that previously relied on these properties should still function as before for the time being, but it is strongly suggested to update these as they are not consistent or frame-rate independent, they may be removed in future.

The internal standard base time unit remains as `16.666ms` for the time being, rather than `1sec` as used in SI. If this were to change later, `1ms` would seem the natural choice. For most users this is not usually necessary to be aware of. 

All examples have been updated to also be frame-rate independent but using `1sec` as their time unit to be more in line with expectations, they are also using the new `updateVelocity` argument in e.g. `Body.setPosition` which makes it easier to correctly control bodies.

### Requesting reviewers, testers and comments

If anybody has any time to help me test this update out further and feedback any issues, it would be greatly appreciated:

- try experiment with the standard examples in different environments (browsers, devices)
  - consider friction, air friction, restitution, sleeping, collisions, constraints
- compare these to using the previous version of Matter.js (e.g. see the [demo](http://brm.io/matter-js/demo/)), results should be very similar or better
- try this branch with your code base (e.g. use the [alpha build on this branch](https://github.com/liabru/matter-js/blob/timing-improve/build/matter.alpha.js))
- try using different `delta` values in `Engine.update` (more than `33.333ms` won't be stable)
- try using `engine.timing.timeScale` (as another way of modifying delta)
- try out the new `updateVelocity`, body velocity and speed setters / getters

### Changes

- Added readonly body.deltaTime
- Added delta property to engine update event
- Added delta argument to various internal functions
- Changed timeScale argument to use delta instead on various internal functions
- Fixed issues when using an engine update delta of 0
- Improved time independence for friction, air friction, restitution, sleeping, collisions, constraints
- Removed optional correction argument from Engine.update
- Removed correction and timeScale from Body.update and Matter.Runner
- Added updateVelocity argument to Body.setPosition, Body.setAngle, Body.translate, Body.rotate
- Added Body.setSpeed, Body.setAngularSpeed
- Added Body.getSpeed, Body.getVelocity, Body.getAngularVelocity
- Changed all velocity functions to be time independent
- Changed examples to be delta independent
